provider/cloudflare: Add proxied option

xla opened the pull request:
This change adds the support for the proxied configuration option for a record which enables origin protection for CloudFlare records.

In order to do so the golang library needed to be changed as the old did not support the option and was using and outdated API version.

There are a couple of open issues which ask for this (#5049, #3805).

***
I was uncertain about the dependency changes and how that is reflected. How are they managed and how can they be changed and bumped?

phinze commented on builtin/providers/cloudflare/resource_cloudflare_record.go:
```diff
@@ -44,15 +47,25 @@ func resourceCloudFlareRecord() *schema.Resource {
+
+			"proxied": &schema.Schema{
+				Type:     schema.TypeBool,
 				Optional: true,
```
Generally for bools we just reflect the upstream default w/ `Default: false`, then below instead of checking `GetOk` you can simply set `d.Get("proxied").(bool)` directly.

phinze commented:
@xla on first review this is looking really nice!

You can find docs about vendoring the new library here: https://github.com/hashicorp/terraform#adding-a-dependency

As for removing the old library - it looks like that has to be done manually by editing the `Godeps/Godeps.json` and removing the path in `vendor/`.

If you like, you can just do the addition of the new library on this PR to get the build passing, and we can take care of the cleanup of the old library on master. :+1:

xla commented:
@phinze Thanks for the quick follow-up. I incorporated your feedback, unfortunately trying to add the dependency results in either errors or a hanging `godep` executable. Maybe you have a reproducible way, so I added you as collaborator in case you find the time to check out the branch.

xla commented:
@phinze Was able to add the missing dependency and the build is green. Unfortunately CloudFlare changed a type returned in the paging part of their API so that this PR is blocked by crackcomm/cloudflare#4.

xla commented:
@phinze Good to go from my side.

stack72 commented:
Tests pass as expected and the deps have been vendored now :)

```
TF_LOG=1 make testacc TEST=./builtin/providers/cloudflare TESTARGS='-run=TestAccCloudFlareRecord' 2>~/tf.log
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/cloudflare -v -run=TestAccCloudFlareRecord -timeout 120m
=== RUN   TestAccCloudFlareRecord_Basic
--- PASS: TestAccCloudFlareRecord_Basic (5.76s)
=== RUN   TestAccCloudFlareRecord_Proxied
--- PASS: TestAccCloudFlareRecord_Proxied (5.24s)
=== RUN   TestAccCloudFlareRecord_Updated
--- PASS: TestAccCloudFlareRecord_Updated (6.29s)
=== RUN   TestAccCloudFlareRecord_forceNewRecord
--- PASS: TestAccCloudFlareRecord_forceNewRecord (7.95s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/cloudflare	25.254s
```

stack72 commented:
Thanks for all the follow up work here @xla :)
